magenta/models/polyphonic_rnn/polyphonic_rnn_lib.py

```diff
@@ -0,0 +1,1377 @@
+  for i in range(len(sequences)):
+    # Try to accomodate for masks...
+    seq = sequences[i]
+    nd = ndim(seq)
+    if nd == 3:
+      pass
+    elif nd < 3:
+      sequences[i] = tf.expand_dims(sequences[i], nd)
+    else:
+      raise ValueError('Ndim too different to correct')
+
+  def check(l):
+    shapes = [shape(s) for s in l]
+    # for now assume -1, can add axis argument later
+    # check shapes match for concatenation
+    compat = [ls for ls in shapes if ls[:-1] == shapes[0][:-1]]
+    if len(compat) != len(shapes):
+      raise ValueError('Tensors *must* be the same dim for now')
+
+  check(sequences)
+  check(outputs_info)
+
+  seqs_shapes = [shape(s) for s in sequences]
+  nd = len(seqs_shapes[0])
+  seq_pack = tf.concat(nd - 1, sequences)
+  outs_shapes = [shape(o) for o in outputs_info]
+  nd = len(outs_shapes[0])
+  init_pack = tf.concat(nd - 1, outputs_info)
+
+  assert len(shape(seq_pack)) == 3
+  assert len(shape(init_pack)) == 2
+
+  def s_e(shps):
+    starts = []
+    ends = []
+    prev_shp = 0
+    for _, shp in enumerate(shps):
+      start = prev_shp
+      end = start + shp[-1]
+      starts.append(start)
+      ends.append(end)
+      prev_shp = end
+    return starts, ends
+
+  # TF puts the initializer in front?
+  def fnwrap(initializer, elems):
+    starts, ends = s_e(seqs_shapes)
+    sliced_elems = [elems[:, start:end] for start, end in zip(starts, ends)]
+    starts, ends = s_e(outs_shapes)
+    sliced_inits = [initializer[:, start:end]
+                    for start, end in zip(starts, ends)]
+    t = []
+    t.extend(sliced_elems)
+    t.extend(sliced_inits)
+    # elems first then inits
+    outs = fn(*t)
+    nd = len(outs_shapes[0])
+    outs_pack = tf.concat(nd - 1, outs)
+    return outs_pack
+
+  r = tf.scan(fnwrap, seq_pack, initializer=init_pack)
+
+  if len(outs_shapes) > 1:
+    starts, ends = s_e(outs_shapes)
+    o = [r[:, :, start:end] for start, end in zip(starts, ends)]
+    return o
+  else:
+    return r
+
+##
+# end metautils
+##
+
+##
+# begin datasets
+##
+
+
+def duration_and_pitch_to_midi(filename, durations, pitches, prime_until=0):
+  # durations and pitches should both be 2D
+  # [time_steps, n_notes]
+
+  sequence = music_pb2.NoteSequence()
+
+  # Hardcode for now, eventually randomize?
+  # or predict...
+  sequence.ticks_per_quarter = mm.STANDARD_PPQ
+  ts = sequence.time_signatures.add()
+  ts.time = 1.0
+  ts.numerator = 4
+  ts.denominator = 4
+
+  ks = sequence.key_signatures.add()
+  ks.key = 0
+  ks.mode = ks.MAJOR
+
+  tempos = sequence.tempos.add()
+  tempos.qpm = 120
+  # ti.simultaneous_notes
+  sn = 4
+  # ti.time_classes
+  # this should be moved out for sure - this is really duration class, pitch!
+  time_classes = [
+      0.125, 0.25, 0.5, 0.75, 1.0, 1.5, 2.0, 3.0, 4.0, 6.0, 8.0, 12.0]
+
+  dt = copy.deepcopy(durations)
+  for n in range(len(time_classes)):
+    dt[dt == n] = time_classes[n]
+
+  # why / 8?
+  delta_times = [dt[..., i] / 8 for i in range(sn)]
```

Comment:
Just to make sure I'm understanding this correctly, does /8 mean that the RNN sees the sequence of notes in 8th note steps?